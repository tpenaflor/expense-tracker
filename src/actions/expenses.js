
// import {v1} from 'uuid'
import db from '../firebase/firebase'

export const addExpensee = (expense) => ({
    type:"ADD_EXPENSE",
    expense
})

export const addExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        const {desc='', note='',amount=0, createAt=0} = expenseData
        const expense = {desc, note, amount, createAt}

        return db.ref(`expenses/${uid}`).push(expense).then((snap) => {
            dispatch(addExpensee({id:snap.key, ...expense}))
        })
    }
}

export const removeExpensee = (id) => ({
    type:"REMOVE_EXPENSE",
    expense : {
        id
    }
})

export const removeExpense = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        return db.ref(`expenses/${uid}/${id}`).remove().then(() =>{
            dispatch(removeExpensee(id))
        })
    }
}

export const editExpensee = (id, update) => ({
    type:"EDIT_EXPENSE",
    id,
    update
})

export const editExpense = (id, update) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        return db.ref(`expenses/${uid}/${id}`).update(update).then(() =>{
            dispatch(editExpensee(id, update))
        })
    }
}

export const setExpensess = (expenses) => ({
    type : 'SET_EXPENSES',
    expenses
})

export const setExpenses = (uid) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        let expenses = []
        return db.ref(`expenses/${uid}`).once('value').then((snap) =>{
            snap.forEach((expense) => {
                expenses.push({id:expense.key, ...expense.val()})
            });

            dispatch(setExpensess(expenses))
        })
    }
}