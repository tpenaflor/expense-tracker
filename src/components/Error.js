import React from 'react'
import {Link} from 'react-router-dom'

const ErrorPage = () => (
    <div>
        404 Page - <Link to="/home">Go Home</Link>
    </div>
)


export default ErrorPage