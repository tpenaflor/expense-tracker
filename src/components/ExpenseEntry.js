import React from 'react'
import {Link} from 'react-router-dom'

import moment from 'moment'

const ExpenseEntry = ({id, desc, amount, createAt}) => (
    <div> 
        <h2>
        <Link to={`/editExpense/${id}`} >{desc}</Link></h2>
        <p>{`${Intl.NumberFormat('en-US',{style:'currency',currency:'USD'}).format(amount/100)} on ${moment(createAt).format("MMM DD, YYYY")}`}</p>
    </div>
)

export default ExpenseEntry