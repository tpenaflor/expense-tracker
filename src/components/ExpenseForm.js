import React from 'react'
import moment from 'moment'

import {SingleDatePicker } from 'react-dates'
import 'react-dates/initialize'

class ExpenseForm extends React.Component {
    state = {
        desc: this.props.desc ? this.props.desc : '',
        note: this.props.note ? this.props.note : '',
        amount: this.props.amount ? (this.props.amount/100).toFixed(2).toString() : '',
        createAt: this.props.createAt ? moment(this.props.createAt) : moment(),
        calFocus : false,
        errorState : ''
    }

    onDescChange = (e) => {
        const desc = e.target.value
        this.setState(()=> ({desc}))
    }

    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(()=> ({note}))
    }

    onAmountChange = (e) => {
        const amount = e.target.value

        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
            return this.setState(()=> ({amount}))
        
    }

    onDateChange = (createAt) => {
        if (createAt)
            this.setState(()=>({createAt}))
    }

    onFocusChange = ({focused : calFocus}) => {
        this.setState(()=>({calFocus}))
    }

    onSubmit = (e) => {
        e.preventDefault();
        
        if (!this.state.desc || !this.state.amount){
            this.setState(()=>({errorState : 'missing desc or amount'}))
        }
        else {
            this.setState(()=>({errorState : ''}))
            this.props.onSubmit({
                desc : this.state.desc,
                note : this.state.note,
                amount : parseFloat(this.state.amount, 10) * 100,
                createAt : this.state.createAt.valueOf()
            })
        }

    }

    render() {
        return (
            <div>
                <form>
                    <input 
                        type='text' 
                        value={this.state.desc} 
                        onChange={this.onDescChange}
                        placeholder='description' 
                        autoFocus />
                    <input 
                        value = {this.state.amount}
                        onChange={this.onAmountChange}
                        type='text' 
                        placeholder='amount' />
                    <SingleDatePicker 
                        date = {this.state.createAt}
                        onDateChange  = {this.onDateChange}
                        focused = {this.state.calFocus}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths = {1}
                        isOutsideRange = {()=> false}
                    />
                    <textarea 
                        value={this.state.note} 
                        onChange={this.onNoteChange}
                        placeholder='add note here (optional)'></textarea>
                    <button onClick={this.onSubmit}>Submit</button>
                    {this.state.errorState && <p>{this.state.errorState}</p>}
                </form>
            </div>
        )
    }
}

export default ExpenseForm