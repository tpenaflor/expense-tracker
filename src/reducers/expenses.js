

const expensesReducerDefaultState = []

const expensesReducer = (state = expensesReducerDefaultState, {type, expense, id, update, expenses}) => {
    switch (type) {
        case 'ADD_EXPENSE' :
            return [...state, expense]
        case 'REMOVE_EXPENSE' :
            return state.filter(({id})=> id !== expense.id)
        case 'EDIT_EXPENSE' :
            return state.map((entry) => {
                if (entry.id === id){
                    return {...entry, ...update}
                }
                return entry
            })
        case 'SET_EXPENSES' :
            return expenses
        default:
            return state
    }
}

export default expensesReducer