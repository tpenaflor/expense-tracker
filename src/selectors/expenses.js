
import moment from 'moment'

export const getVisibleExpenses = (expenses, {text, startDate, endDate, sortBy}) => {

    return expenses.filter(({desc, createAt})=> {
        const createMoment = moment(createAt)
        // const startMatch = typeof startDate !== 'number' || createAt >= startDate
        // const endMatch = typeof endDate !== 'number' || createAt <= endDate
        const startMatch = startDate ? startDate.isSameOrBefore(createMoment, 'day'): true
        const endMatch = endDate ? endDate.isSameOrAfter(createMoment, 'day'): true
        const textMatch = desc.toLowerCase().includes(text.toLowerCase())
        
        return startMatch && endMatch && textMatch
    }).sort((a, b)=>{
        if (sortBy === 'date') return a.createAt - b.createAt
        if (sortBy === 'amount') return a.amount - b.amount
    });
}

export const getTotalExpenses = (expenses) => expenses.reduce((total, {amount}) => total += amount, 0)

export default getVisibleExpenses