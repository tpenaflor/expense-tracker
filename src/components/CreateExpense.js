import React from 'react'
import {connect} from 'react-redux'

import ExpenseForm from './ExpenseForm'
import {addExpense} from '../actions/expenses'

export const CreateExpense = ({onSubmit, history}) => (
    <div>
        <h2>CreateExpense</h2>
        <ExpenseForm 
            onSubmit = {(expense) => {
                onSubmit(expense)
                history.push('/')
            }}
        />
    </div>

)

const mapDispatchToProps = (dispatch)=> ({
    onSubmit : (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(CreateExpense)