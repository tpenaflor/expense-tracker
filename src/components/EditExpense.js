import React from 'react'
import {connect} from 'react-redux'

import ExpenseForm from './ExpenseForm'
import {editExpense, removeExpense} from '../actions/expenses'

export const EditExpensePage = (props) => (
    <div className='content-container'>
        <h2>Edit {props.expense.desc} </h2>
        <ExpenseForm 
            desc = {props.expense.desc}
            amount = {props.expense.amount}
            createAt = {props.expense.createAt}
            note = {props.expense.note}
            onSubmit = {
            (expense) => {
                props.editExpense(props.expense.id, expense)
                props.history.push('/')
            }} 
            />
        
        <button onClick={()=>{
                props.removeExpense(props.expense.id)
                props.history.push('/')
            }}>Remove</button>
    </div>
)

const VerifyEditPage = (props) => (
    <div>
        {props.expense ? <EditExpensePage {...props} /> : props.history.push('/editExpense/')}
    </div>
)

const mapToProps = (state, props) =>
{
    return {
        expense : state.expenses.find((expense) => expense.id === props.match.params.id),
        all : state.expenses
    }
}

const mapDispToProps = (dispatch) => ({
    editExpense : (id,expense) => dispatch(editExpense(id, expense)),
    removeExpense : (id) => dispatch(removeExpense(id))
})

export default connect(mapToProps,mapDispToProps)(VerifyEditPage)