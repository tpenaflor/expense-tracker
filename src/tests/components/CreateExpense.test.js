import React from 'react'
import {shallow} from 'enzyme'
import {CreateExpense} from '../../components/CreateExpense'
import ExpenseForm from '../../components/ExpenseForm'
import expenses from '../fixtures/expenses'

let onSubmit, history, wrapper;
beforeEach( ()=> {
    onSubmit =  jest.fn()
    history = {push:jest.fn()}
    wrapper = shallow(<CreateExpense onSubmit={onSubmit} history={history} />)
})

afterEach( () => {
    expect(wrapper).toMatchSnapshot()

})

test('test create expense page', ()=> {
    
})

test('test create expense submit', ()=> {
    wrapper.find(ExpenseForm).prop('onSubmit')(expenses[0])
        
    expect(onSubmit).toHaveBeenLastCalledWith(expenses[0])
    expect(history.push).toHaveBeenLastCalledWith('/')
})
