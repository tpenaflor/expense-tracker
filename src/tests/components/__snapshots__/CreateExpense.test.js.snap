// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test create expense page 1`] = `
<div
  className="content-container"
>
  <h2>
    CreateExpense
  </h2>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;

exports[`test create expense submit 1`] = `
<div
  className="content-container"
>
  <h2>
    CreateExpense
  </h2>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
